File: backend/app/schemas/wastage.py
from typing import Optional
from uuid import UUID
from datetime import date, datetime

from pydantic import BaseModel


class WastageCreate(BaseModel):
    source_type: str = "store"
    store_id: Optional[UUID] = None
    product_id: Optional[UUID] = None
    ingredient_id: Optional[UUID] = None
    date: date
    quantity: int
    reason: str
    notes: Optional[str] = None


class WastageResponse(BaseModel):
    id: UUID
    source_type: str
    store_id: Optional[UUID] = None
    store_name: Optional[str] = None
    product_id: Optional[UUID] = None
    product_name: Optional[str] = None
    product_unit: Optional[str] = None
    ingredient_id: Optional[UUID] = None
    ingredient_name: Optional[str] = None
    ingredient_unit: Optional[str] = None
    date: date
    quantity: int
    unit_price: Optional[float] = None
    total_price: Optional[float] = None
    unit_cost_snapshot: Optional[float] = None
    total_cost_snapshot: Optional[float] = None
    cost_source: Optional[str] = None
    is_estimated_cost: bool = False
    reason: str
    notes: Optional[str] = None
    recorded_by: Optional[UUID] = None
    recorded_by_name: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/wastage.ts
import apiClient from './client';

export interface WastageDetail {
  id: string;
  source_type: 'store' | 'production';
  store_id: string | null;
  store_name: string | null;
  product_id: string | null;
  product_name: string | null;
  product_unit: string | null;
  ingredient_id: string | null;
  ingredient_name: string | null;
  ingredient_unit: string | null;
  date: string;
  quantity: number;
  unit_price: number | null;
  total_price: number | null;
  unit_cost_snapshot: number | null;
  total_cost_snapshot: number | null;
  cost_source: string | null;
  is_estimated_cost: boolean;
  reason: string;
  notes: string | null;
  recorded_by: string | null;
  recorded_by_name: string | null;
  created_at: string;
}

export interface CreateWastagePayload {
  source_type: 'store' | 'production';
  store_id?: string;
  product_id?: string;
  ingredient_id?: string;
  date: string;
  quantity: number;
  reason: string;
  notes?: string;
}

export const wastageApi = {
  list: async (params?: {
    skip?: number;
    limit?: number;
    source_type?: 'store' | 'production';
    store_id?: string;
    product_id?: string;
    ingredient_id?: string;
  }): Promise<WastageDetail[]> => {
    const response = await apiClient.get('/wastage/', { params });
    return response.data;
  },

  create: async (data: CreateWastagePayload): Promise<WastageDetail> => {
    const response = await apiClient.post('/wastage/', data);
    return response.data;
  },
};
